### ISSUE TITLE
Questions / suggestions on how warnings are handled / emitted in nipype

### ISSUE THREAD

**Issue by bcipolli** [nipy/nipype#1245]
I don't understand some of the use of warnings in `nipype`. I made some changes while working on #1221 to manage things, and thought it's useful to discuss.

* I find some inconsistency in raising warnings vs. Errors:
    * Some interface errors cause errors (if two conflicting trait values are set,  in `BaseTraitedSpec._xor_warn)
    * others cause warnings (if a required trait is not set, in `BaseTraitedSpec._requires_warn`).
    * **Suggested solution**: Raise an error if a required trait is not set.

* Some modules explicitly set `UserWarning` to `always`.
    * Why do this? By default they are set to `always`, so if they're not it's because someone doesn't want them on!
    * Why not raise important warnings as errors?
    * If this is really desired, why not do this in `nipype.__init__` once, for all modules (and limit it to `nipype` by setting `module='nipype'` in `filterwarnings`?
    * **Suggested solution**: raise errors where it really matters, and delete the `filterwarnings` lines.

* The test output is *really* hard to read due to all the warnings
    * If warnings are so important, shouldn't they be explicitly caught in the tests?
    * For unimportant warnings, can we add a filter to remove them (either in `.travis.yml` or in a test runner or in `fixes/numpy/testing/nosetester.py`?
    * **Suggested solution**: use [`warnings.catch_warnings`](https://docs.python.org/2/library/warnings.html#testing-warnings) to test warnings; suppress verbose, unneded warnings (i.e. the FSL warnings) in `.travis.yml`

**Comment by mwaskom** [comment]
> others cause warnings (if a required trait is not set, in BaseTraitedSpec._requires_warn).
> Suggested solution: Raise an error if a required trait is not set.

I think the problem here is that the warning gets fired as the *first* trait gets set, but it's possible that you set the trait it requires later (e.g. during a connection). Whereas with xor traits, you can raise on the *second* trait, at which point you are certain that something has been done wrong.

With that said, I generally think warnings are noise so I'm all for reducing them. I have never been a fan of the requires warning because it is so often wrong.

**Comment by oesteban** [comment]
I agree with @bcipolli, we should revise warnings and info. Some related issues are #1106 and #1143.
